photosynthesis_metrics/gmsd.py

```diff
@@ -0,0 +1,253 @@
+        target = F.avg_pool2d(target, kernel_size=2, stride=2, padding=padding)
+        
+        score = _gmsd(
+            prediction, target, reduction=self.reduction)
+
+        return score
+    
+
+class MultiScaleGMSDLoss(_Loss):
+    r"""Creates a criterion that measures the structural similarity index error between
+    each element in the input :math:`x` and target :math:`y`.
+
+
+    Args:
+        reduction: Specifies the reduction to apply to the output:
+            ``'none'`` | ``'mean'`` | ``'sum'``. ``'none'``: no reduction will be applied,
+            ``'mean'``: the sum of the output will be divided by the number of
+            elements in the output, ``'sum'``: the output will be summed. Default: ``'mean'``
+        data_range: The difference between the maximum and minimum of the pixel value,
+            i.e., if for image x it holds min(x) = 0 and max(x) = 1, then data_range = 1.
+            The pixel value interval of both input and output should remain the same.
+        scale_weights: Weights for different scales. Must contain 4 floating point values.
+        chromatic: Flag to use MS-GMSDc algorithm from paper.
+            It also evaluates chromatic components of the image. Default: True
+        beta1: Algorithm parameter. Weight of chromatic component in the loss.
+        beta2: Algorithm parameter. Small constant, see [1].
+        beta3: Algorithm parameter. Small constant, see [1].
+        
+    Reference:
+        [1] GRADIENT MAGNITUDE SIMILARITY DEVIATION ON MULTIPLE SCALES (2017)
+            http://www.cse.ust.hk/~psander/docs/gradsim.pdf
+    """
+
+    def __init__(self, reduction: str = 'mean', data_range: Union[int, float] = 1.,
+                 scale_weights: Optional[Union[Tuple[float], List[float]]] = None,
+                 chromatic: bool = False, beta1: float = 0.01, beta2: float = 0.32,
+                 beta3: float = 15.) -> None:
+        super().__init__()
+
+        # Generic loss parameters.
+        self.reduction = reduction
+
+        # Loss-specific parameters.
+        self.data_range = data_range
+        
+        # Values from the paper
+        if scale_weights is None:
+            self.scale_weights = torch.tensor([0.096, 0.596, 0.289, 0.019])
+        else:
+            self.scale_weights = torch.tensor(scale_weights)
+
+        self.chromatic = chromatic
+        self.beta1 = beta1
+        self.beta2 = beta2
+        self.beta3 = beta3
+            
+    def forward(self, prediction: torch.Tensor, target: torch.Tensor) -> torch.Tensor:
+        r"""Computation of Gradient Magnitude Similarity Deviation (GMSD) as a loss function.
+
+        Args:
+            prediction: Tensor of prediction of the network.
+            target: Reference tensor.
+
+        Returns:
+            Value of GMSD loss to be minimized. 0 <= GMSD loss <= 1.
+        """
+        _validate_input(x=prediction, y=target, scale_weights=None)
```

Comment:
```suggestion
        _validate_input(x=prediction, y=target, scale_weights=self.scale_weights)
```

Reply:
We have different number of weights in MS-SSIM and MS-GMSD (5 and 4 respectively). So this change leads to assertion errors in `_validate_input` function.

Reply:
`_validate_input` will support flexible number of weights after #55. It's already implemented.

Reply:
May user use custom weights of different size? like `[0.4, 0.5, 0.3, 0.2, 0.1]`?

Reply:
Weights should sum to 1. A added weights normalization in `__init__`, so even when user writes `[2., 4., 1., 1.]` everything should work fine.

Reply:
The point is to use 5 weights instead of 4. Let's make them [0.2, 0.2, 0.2, 0.2, 0.2] or [0.3, 0.3, 0.3].